server/utils/APIWrapper.ts

```diff
@@ -1,33 +1,93 @@
+import Ajv, { ErrorObject, JTDSchemaType } from "ajv/dist/jtd";
 import mongoose from "mongoose";
 import { NextApiResponse } from "next";
 import { getSession } from "next-auth/client";
 import {
-  HttpMethod,
   InternalRequest,
   InternalResponseData,
   Role,
+  StringsDict,
+  UnknownDict,
 } from "src/utils/types";
 
-interface RouteConfig {
+interface RouteConfig<Q_P, B> {
   requireSession: boolean;
   roles?: Array<Role>;
   handleResponse?: boolean; // handleResponse if the route handles setting status code and body
+  bodySchema?: JTDSchemaType<B>;
+  querySchema?: JTDSchemaType<Q_P>;
 }
 
-interface Route<T> {
-  config?: RouteConfig;
+export interface Route<Q_P, B, R> {
+  config?: RouteConfig<Q_P, B>;
   handler: (
-    req: InternalRequest,
-    res: NextApiResponse<InternalResponseData<T>>
-  ) => Promise<T>;
+    req: InternalRequest<Q_P, B>,
+    res: NextApiResponse<InternalResponseData<R>>
+  ) => Promise<R>;
 }
 
-function APIWrapper(
-  routeHandlers: Partial<Record<HttpMethod, Route<unknown>>>
+export type UnvalidatedInputRoute = {
+  config: Omit<
+    RouteConfig<StringsDict, UnknownDict>,
+    "bodySchema" | "querySchema"
+  >;
+  handler: (
+    req: InternalRequest<StringsDict, UnknownDict>,
+    res: NextApiResponse
+  ) => Promise<unknown>;
+};
+
+/**
+ * @deprecated Please import { APIWrapper } instead, which enforces schemas
+ * Does not validate input schemas for the API
+ * @param routeHandlers
+ */
+export function UnvalidatedInputAPIWrapper(
+  routeHandlers: Partial<
+    Record<"GET" | "PATCH" | "POST" | "PUT", UnvalidatedInputRoute>
+  >
+) {
+  return APIWrapper(routeHandlers);
+}
+
+/**
+ * In order to use the body or query params from the API wrapper, you must provide a schema
+ * to validate the input
+ * See  src/pages/api/applications/[id]/chapter.ts and
+ * src/pages/api/applications/nonprofit.ts
+ * as references
+ * @param routeHandlers
+ * @constructor
+ */
+export function APIWrapper<
+  GET_Q_P,
```

Comment:
Very weird stuff to get the types to play along nicely. I tried many variants of `Partial<Record<HTTPMethod, Route<unknown, unknown, unknown>>`, and it refused to work

Reply:
Loool gotta love Typescript.  I'll think about this param situation a little more later to see if it can be reduced a little.